### ISSUE TITLE
Steps to compiling on Apple Silicon machines, and request for official support

### ISSUE THREAD

**Issue by leo-maxwell** [TypesettingTools/Aegisub#191]
I managed to find an easy way to compile a fully native Apple Silicon package and it's very simple to follow. I suggest it's time to add support for Apple Silicon Macs, and it benefits a lot of subtitle workers.

**I would assume that everybody uses Homebrew as their package manager. The operations described are all done in a case-sensitive APFS partition on my Apple Silicon computer.**

After all the dependency being installed, first the path in the FLAGS should change for Apple Silicon homebrew packages. That is

```
export LDFLAGS="-L/opt/homebrew/opt/icu4c/lib"
export CPPFLAGS="-I/opt/homebrew/opt/icu4c/include"
export PKG_CONFIG_PATH="/opt/homebrew/opt/icu4c/lib/pkgconfig"
```

And when compiling, in `src/dialog_colorpicker.cpp`, line 400 and line 401, the type `NSUInteger` is somehow broken, and here is the error message

```
../src/dialog_colorpicker.cpp:400:2: error: unknown type name 'NSUInteger'
        NSUInteger width = CGImageGetWidth(img);
        ^
../src/dialog_colorpicker.cpp:401:2: error: unknown type name 'NSUInteger'
        NSUInteger height = CGImageGetHeight(img);
        ^
```

From [Apple Developer site](https://developer.apple.com/documentation/objectivec/nsuinteger), the `NSUInteger` is simply `unsigned long`, and I replaced these two `NSUInteger` to `unsigned long`, it worked fine.

After following steps in the readme, the last problem is the codesigning. This problem is particularly annoying for open source communities but it has an easy fix. I just executed the command

```
codesign -s - -f -vvvvv --deep /Volumes/Develop/Aegisub/build_static/Aegisub.app
```

And it worked well, you can double click it and it launched right away. From the Activity Monitor, it is a fully native software. But it is too early to claim victory, because the language of the software interface is a mixture of English and Japanese. But my computer's default language is not English nor Japanese. Also when performing the test command, one if the tests failed and it was

```
[ RUN      ] lagi_iconv.Fallbacks
../tests/tests/iconv.cpp:43: Failure
Expected: fallback.Convert("\\") doesn't throw an exception.
  Actual: it throws.
```

I think this problem is related to the language issue. In short, I suggest that these changes should be made to achieve the goal of adding support for Apple Silicon machines:

- Add the section in the readme file to guide people how to compile on Apple Silicon machines
- Modify the `src/dialog_colorpicker.cpp` file to replace the `NSUInteger`, or find a better way to solve this problem
- Modify `tools/osx-fix-libs.py` to add an extra step of re-signing all the fixed libs
- Investigate the language problem
- In Github Actions you could use Apple Silicon machines to automatically add an Apple Silicon package at each release

Thank you for your reading and I sincerely hope that such a good open source software can be fully supported on newer machines.

**Comment by arch1t3cht** [comment]
For what it's worth, compilation worked out of the box for me on my arm64 mac machine, except for that one failing test and the code signing when bundling (neither of which are relevant when simply running the built binary). That's why none of these issues where addressed so far. I couldn't reproduce the `NSUInteger` issue or the language one.

> In Github Actions you could use Apple Silicon machines to automatically add an Apple Silicon package at each release

To my knowledge there are no free Apple Silicon runners available on GitHub.

I/We can still address these issues when we can, I just want to make clear that Apple Silicon wasn't being ignored up until now or anything.

**Comment by shinra-electric** [comment]
> To my knowledge there are no free Apple Silicon runners available on GitHub.

Just an FYI, Apple Silicon runners are [available for free](https://github.blog/changelog/2024-01-30-github-actions-introducing-the-new-m1-macos-runner-available-to-open-source) on GitHub now.

**Comment by leo-maxwell** [comment]
Thank you for telling me this important information! I have already set up a repo that uses GitHub Actions to build Apple Silicon native packages and released a binary. If you are interested, please go to: 

https://github.com/leo-maxwell/Aegisub_Apple_Silicon_build

Some necessary modifications are made to make this happen. The `NSUInteger` problem persists and I simply defined that symbol to `size_t`. For other modifications, please check the repo.

**Comment by shinra-electric** [comment]
>I have already set up a repo that uses GitHub Actions to build Apple Silicon native packages and released a binary. 

Thanks for setting that up. I always found Aegisub finicky to build...

I was going to open an issue related to the build, but you seem to have them turned off...? 

Basically, the user interface is not using my primary language (English - Ireland), but is skipping it and going to my third language preference (Japanese), also skipping my second preference which is simplified Chinese. 

<img width="406" alt="Screenshot 2024-04-17 at 20 36 55" src="https://github.com/TypesettingTools/Aegisub/assets/50119606/7a41f7d1-e839-40fd-a5b6-b0a36c44a291">

For comparison, the build I downloaded from [here](https://gist.github.com/frozenpandaman/92a84cd2ba690bb2399afd04e695bf8c) by rcombs (1098 - 23/12/21) will default to English.

<img width="386" alt="Screenshot 2024-04-17 at 20 37 24" src="https://github.com/TypesettingTools/Aegisub/assets/50119606/7fd78c2b-e66e-4625-8ba2-1430977a00c3">

**Comment by leo-maxwell** [comment]
That's a problem that I met myself when testing the local build. I can read Chinese and if you are familiar with Chinese characters (or kanji, in Japanese), I think you will find the button on the menubar that reads "settings" and change the language to English. Then the problem is fixed, and the settings should be written to the Application Support folder. Of course, this is an inconvenience and I will create an issue for that in my repo. Sorry for not enabling the issue page and that problem is fixed now. This issue is closed now for good and maybe I will contribute to this repo someday in the future.

**Comment by lrtfm** [comment]
Installing libiconv with Homebrew and updating the CPPFLAGS and LDFLAGS paths seems to fix the test failures from the meson test -C build_static --verbose command. It looks like libiconv wasn't really on the system. see the discussion here: [libiconv missing when compiling on macOS](https://github.com/OpenModelica/OpenModelica/issues/10944)

Following [leo-maxwell](https://github.com/leo-maxwell)'s, here's a quick rundown of the steps:

1. Patch the files (the patch file: [aegisub-osx.patch](https://github.com/user-attachments/files/15748777/aegisub-osx.patch))


    1. Fix the issue with `NSUInteger`

        ```
        diff --git a/src/dialog_colorpicker.cpp b/src/dialog_colorpicker.cpp
        index e0d8882de..1af98dfa2 100644
        --- a/src/dialog_colorpicker.cpp
        +++ b/src/dialog_colorpicker.cpp
        @@ -397,8 +397,8 @@ void ColorPickerScreenDropper::DropFromScreenXY(int x, int y) {
                CGGetDisplaysWithPoint(CGPointMake(x, y), 1, &display_id, &display_count);

                agi::scoped_holder<CGImageRef> img(CGDisplayCreateImageForRect(display_id, CGRectMake(x - resx / 2, y - resy / 2, resx, resy)), CGImageRelease);
        -       size_t width = CGImageGetWidth(img);
        -       size_t height = CGImageGetHeight(img);
        +       NSUInteger width = CGImageGetWidth(img);
        +       NSUInteger height = CGImageGetHeight(img);
                std::vector<uint8_t> imgdata(height * width * 4);

                agi::scoped_holder<CGColorSpaceRef> colorspace(CGColorSpaceCreateDeviceRGB(), CGColorSpaceRelease);
        ```

    2. In my opinion, you don't need to use sudo when fixing the libs.

        ```
        diff --git a/tools/osx-bundle.sh b/tools/osx-bundle.sh
        index bb8535f4f..ac0f58080 100755
        --- a/tools/osx-bundle.sh
        +++ b/tools/osx-bundle.sh
        @@ -103,7 +103,7 @@ mkdir -vp "${PKG_DIR}/Contents/Resources/en.lproj"

         echo
         echo "---- Fixing libraries ----"
        -python3 "${SRC_DIR}/tools/osx-fix-libs.py" "${PKG_DIR}/Contents/MacOS/aegisub" || exit $?
        +sudo python3 "${SRC_DIR}/tools/osx-fix-libs.py" "${PKG_DIR}/Contents/MacOS/aegisub" || exit $?

         echo
         echo "Done creating \"${PKG_DIR}\""
        ```

    3. `bless` seems not needed for new system

        ```
        diff --git a/tools/osx-dmg.sh b/tools/osx-dmg.sh
        index 15153e191..24a9ec64b 100755
        --- a/tools/osx-dmg.sh
        +++ b/tools/osx-dmg.sh
        @@ -57,9 +57,9 @@ echo "---- Mounting image ----"
         DEV_NAME=`/usr/bin/hdiutil attach -readwrite -noverify -noautoopen "${DMG_RW_PATH}" |awk '/GUID_partition_scheme/ {print $1}'`
         echo "Device name: ${DEV_NAME}"

        -# echo
        -# echo "---- Setting bless -openfolder ----"
        -# bless -openfolder "/Volumes/${PKG_NAME_VOLUME}"
        +echo
        +echo "---- Setting bless -openfolder ----"
        +bless -openfolder "/Volumes/${PKG_NAME_VOLUME}"

         echo
         echo "---- Setting root icon using SetFile ----"
        ```

5. Install dependencies using homebrew.

    ```
    brew install boost cmake ffmpeg ffms2 fftw \
        hunspell libass libiconv luarocks \
        meson ninja openal-soft pkg-config \
        portaudio pulseaudio uchardet wxwidgets zlib
    ```

6. Set up the environment.

    ```
    export LDFLAGS="-L/opt/homebrew/opt/libiconv/lib"
    export CPPFLAGS="-I/opt/homebrew/opt/libiconv/include"

    export PKG_CONFIG_PATH="/opt/homebrew/opt/icu4c/lib/pkgconfig:/opt/homebrew/opt/openal-soft/lib/pkgconfig"
    ```

7. Configure and build (with codesign).

    ```
    meson build_static -Ddefault_library=static -Dbuildtype=debugoptimized -Dbuild_osx_bundle=true -Dlocal_boost=true
    meson compile -C build_static
    meson test -C build_static --verbose
    meson compile osx-bundle -C build_static
    codesign --sign - --force --deep build_static/Aegisub.app
    meson compile osx-build-dmg -C build_static
    ```

**Comment by leo-maxwell** [comment]
Thank you for your advice on addressing the language issue! But unfortunately, it did not solve the issue...I tested locally with no luck, but I appreciate your engagement!

**Comment by leo-maxwell** [comment]
Although the test has failed, the produced app runs fine without the previous issue. You can see the details in my repo's Actions page about why it failed. Thank you for the suggestion! It definitely solved problems, just not all of them.

**Comment by lrtfm** [comment]
The variables `LDFLAGS`, `CPPFLAGS`, and `PKG_CONFIG_PATH` should be set in section `Configure` ([ci.yml  L46-L48](https://github.com/leo-maxwell/Aegisub_Apple_Silicon_build/blob/b7b99f1955675439e2f1aeb3c639d58b285e1a47/.github/workflows/ci.yml#L46C1-L48C1))

```
      - name: Configure
        run: |
          export LDFLAGS="-L/opt/homebrew/opt/libiconv/lib"
          export CPPFLAGS="-I/opt/homebrew/opt/libiconv/include"
          export PKG_CONFIG_PATH="/opt/homebrew/opt/icu4c/lib/pkgconfig:/opt/homebrew/opt/openal-soft/lib/pkgconfig" 
          meson build_static -Ddefault_library=static -Dbuildtype=debugoptimized -Dbuild_osx_bundle=true -Dlocal_boost=true
```

**Comment by RadonCRM** [comment]
To build a working dmg for 3.4.0, I had to set `MACOSX_DEPLOYMENT_TARGET` to `14.0` (suggested by arch1t3cht to work around the CGDisplayCreateImageForRect issue) and follow steps 2-4 from lrtfm's instructions above. I also did this (from the project's readme.md):

```
export LDFLAGS="-L/usr/local/opt/icu4c/lib"
export CPPFLAGS="-I/usr/local/opt/icu4c/include"
export PKG_CONFIG_PATH="/usr/local/opt/icu4c/lib/pkgconfig"
```

Without doing all of the above, despite getting the dmg to build and being able to mount it, the executable would crash on startup and also had a signing issue according to WhatsYourSign. 
After following the steps, the program and the dmg are working as intended and also marked as validly signed.
